src/gui/plugins/insert_model/InsertModel.cc

```diff
@@ -0,0 +1,221 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#include <ignition/msgs/boolean.pb.h>
+#include <ignition/msgs/stringmsg.pb.h>
+
+#include <sdf/Root.hh>
+#include <sdf/parser.hh>
+
+#include <ignition/common/Console.hh>
+#include <ignition/common/Profiler.hh>
+#include <ignition/common/Filesystem.hh>
+#include <ignition/gui/Application.hh>
+#include <ignition/gui/MainWindow.hh>
+#include <ignition/plugin/Register.hh>
+#include <ignition/transport/Node.hh>
+#include <ignition/transport/Publisher.hh>
+
+#include "ignition/gazebo/EntityComponentManager.hh"
+#include "ignition/gazebo/gui/GuiEvents.hh"
+
+#include "InsertModel.hh"
+
+namespace ignition::gazebo
+{
+  class InsertModelPrivate
+  {
+    /// \brief Ignition communication node.
+    public: transport::Node node;
+
+    /// \brief Mutex to protect mode
+    public: std::mutex mutex;
+
+    /// \brief Transform control service name
+    public: std::string service;
+
+    /// \brief The grid model that the qml gridview reflects
+    public: GridModel gridModel;
+  };
+}
+
+using namespace ignition;
+using namespace gazebo;
+
+/////////////////////////////////////////////////
+GridModel::GridModel() : QStandardItemModel()
+{
+}
+
+/////////////////////////////////////////////////
+void GridModel::AddLocalModel(LocalModel &_model)
+{
+  IGN_PROFILE_THREAD_NAME("Qt thread");
+  IGN_PROFILE("GridModel::AddLocalModel");
+  QStandardItem *parentItem{nullptr};
+
+  parentItem = this->invisibleRootItem();
+
+  auto localModel = new QStandardItem(QString::fromStdString(_model.name));
+  localModel->setData(QString::fromStdString(_model.thumbnailPath),
+                      this->roleNames().key("thumbnail"));
+  localModel->setData(QString::fromStdString(_model.name),
+                      this->roleNames().key("name"));
+  localModel->setData(QString::fromStdString(_model.sdfPath),
+                      this->roleNames().key("sdf"));
+
+  parentItem->appendRow(localModel);
+}
+
+/////////////////////////////////////////////////
+QHash<int, QByteArray> GridModel::roleNames() const
+{
+  return {
+    std::pair(100, "thumbnail"),
+    std::pair(101, "name"),
+    std::pair(102, "sdf"),
+  };
+}
+
+/////////////////////////////////////////////////
+InsertModel::InsertModel()
+  : ignition::gui::Plugin(),
+  dataPtr(std::make_unique<InsertModelPrivate>())
+{
+  ignition::gui::App()->Engine()->rootContext()->setContextProperty(
+      "LocalModelList", &this->dataPtr->gridModel);
+}
+
+/////////////////////////////////////////////////
+InsertModel::~InsertModel() = default;
+
+/////////////////////////////////////////////////
+void InsertModel::FindLocalModels(const std::string &_path)
+{
+  std::string path = _path;
+  // Recurse if directory
+  if (common::isDirectory(path))
```

Comment:
In general it's good to recurse into the directories like this. However in this case, this can result in displaying resources that aren't available.

For example, if `IGN_GAZEBO_RESOURCE_PATH=/tmp/models`, and that folder structure is:

```
└── models
    ├── modelA
    └── subfolder
             └── modelB
``` 

This will work:

```
model://modelA
```

But this won't work:

```
model://modelB
```

So I think we should only display the top-level models.

Reply:
f07d872f88652a5600dfcb8a9419a5bb0156461f